File: spec/json/bloomfilter_spec.rb
require "spec_helper"

describe JsonBloomfilter do

  describe "#initialize" do
    it "should take the appropriate options" do
      seed = Time.now.to_i - 24*60*60
      bf = JsonBloomfilter.new size: 200, hashes: 10, seed: seed
      expect(bf.to_hash["size"]).to be == 200
      expect(bf.to_hash["hashes"]).to be == 10
      expect(bf.to_hash["seed"]).to be == seed
    end

    it "should be initializable with a field serialized by another bloom filter" do
      bf1 = JsonBloomfilter.new
      bf1.add "foo"
      bf2 = JsonBloomfilter.new bf1.to_hash
      expect(bf2.test "foo").to be_true
    end
  end

  context "with an instance" do
    before :each do
      @bf = JsonBloomfilter.new
      @bf.add("foobar")
    end

    describe "#add, #test" do
      it "should add a key" do
        expect(@bf.test "foo").to be_false
        @bf.add "foo"
        expect(@bf.test "foo").to be_true
      end

      it "should not change anything if added twice" do
        expect(@bf.test "foobar").to be_true
        bits = @bf.to_hash["bits"]
        @bf.add "foobar"
        expect(@bf.test "foobar").to be_true
        expect(@bf.to_hash["bits"]).to be == bits
      end
    end

    describe "#clear" do
      it "should clear the bit array" do
        expect(@bf.to_hash["bits"]).not_to be == [0,0,0,0]
        @bf.clear
        expect(@bf.to_hash["bits"]).to be == [0,0,0,0]
      end
    end

    describe "#to_hash" do
      it "should return the serialisable hash" do
        hash = @bf.to_hash
        expect(hash).to be_a(Hash)

        expect(hash).to have_key("seed")
        expect(hash["seed"]).to be_a(Integer)

        expect(hash).to have_key("hashes")
        expect(hash["hashes"]).to be_a(Integer)

        expect(hash).to have_key("size")
        expect(hash["size"]).to be_a(Integer)

        expect(hash).to have_key("bits")
        expect(hash["bits"]).to be_a(Array)
      end
    end

    describe "#to_json" do
      it "should return the hash serialised" do
        expect(@bf.to_json).to be == JSON.generate(@bf.to_hash)
      end
    end

  end
end

File: spec/javascripts/bloomfilter_spec.js
describe("JsonBloomfilter", function() {

  describe(".build", function() {
    it("should generate a BloomFilter with the right number of hashes and size", function() {
      bf = JsonBloomfilter.build(1000, 0.01);
      expect(bf.toHash()["hashes"]).toBe(7);
      expect(bf.toHash()["size"]).toBe(9586);
    });

    it("should optionally take an array of strings instead of a capacity", function() {
      bf = JsonBloomfilter.build(["foo", "bar"], 0.01);
      expect(bf.toHash()["hashes"]).toBe(7);
      expect(bf.toHash()["size"]).toBe(20);
    });

    it("should require a positive integer capacity", function() {
      expect(function(){new JsonBloomfilter.build(0, 0.01)}).toThrow("Capacity needs to be a positive integer")
    });

  });

  describe("#initialize", function() {
    it("should take the appropriate options", function() {
      seed = (new Date().getTime()/1000) - 24*60*60;
      bf = new JsonBloomfilter({ size: 200, hashes: 10, seed: seed });
      expect(bf.toHash()["size"]).toBe(200);
      expect(bf.toHash()["hashes"]).toBe(10);
      expect(bf.toHash()["seed"]).toBe(seed);
    });

    it("should be initializable with a field serialized by another bloom filter", function() {
      bf1 = new JsonBloomfilter();
      bf1.add("foo");
      bf2 = new JsonBloomfilter(bf1.toHash());
      expect(bf2.test("foo")).toBe(true);
    });

    it("should initialize the field with the right size", function() {
      bf = new JsonBloomfilter({size: 100});
      expect(bf.toHash()["bits"].length).toBe(4);
    });
  });

  describe("with an instance", function() {
    var bf;

    beforeEach(function() {
      bf = new JsonBloomfilter();
      bf.add("foobar");
    });

    describe("#add, #test", function() {
      it("should add a key", function() {
        expect(bf.test("foo")).toBe(false);
        bf.add("foo");
        expect(bf.test("foo")).toBe(true);
      });

      it("should be able to add and test more than one key at a time", function() {
        expect(bf.test("foo")).toBe(false);
        expect(bf.test("bar")).toBe(false);
        bf.add(["foo", "bar"]);
        expect(bf.test(["foo", "bar"])).toBe(true);
      });

      it("should not change anything if added twice", function() {
        expect(bf.test("foobar")).toBe(true);
        bits = bf.toHash()["bits"];
        bf.add("foobar");
        expect(bf.test("foobar")).toBe(true);
      });
    });

    describe("#clear", function() {
      it("should clear the bit array", function() {
        expect(bf.bits.toString()).not.toBe("0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000");
        bf.clear();
        expect(bf.bits.toString()).toBe("0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000");
      });
    });

    describe("#toHash", function() {
      it("should return the serialisable hash", function() {
        hash = bf.toHash();
        expect(hash instanceof Object).toBe(true);
        expect(hash["seed"]).not.toBeUndefined();
        expect(hash["hashes"]).not.toBeUndefined();
        expect(hash["size"]).not.toBeUndefined();
        expect(hash["bits"]).not.toBeUndefined();
      });
    });

    describe("#toJson", function() {
      it("should return the hash serialised", function() {
        expect(bf.toJson()).toBe(JSON.stringify(bf.toHash()));
      });
    });

  });
});